src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs

```diff
@@ -6,22 +6,32 @@
 namespace System.Text.RegularExpressions
 {
     [Serializable]
-    internal sealed class RegexParseException : ArgumentException
+    public sealed class RegexParseException : ArgumentException
     {
-        private readonly RegexParseError _error; // tests access this via private reflection
-
         /// <summary>Gets the error that happened during parsing.</summary>
-        public RegexParseError Error => _error;
+        public RegexParseError Error { get; }
 
         /// <summary>Gets the offset in the supplied pattern.</summary>
         public int Offset { get; }
 
-        public RegexParseException(RegexParseError error, int offset, string message) : base(message)
+        public RegexParseException(RegexParseError error, int offset)
+        {
```

Comment:
Does our calling code always include both error and offset in the message and would it be better for this exception to take care of constructing that message for them?

Reply:
The `message` cannot be manipulated by a public ctor, that's on purpose. The ctor with the `message` parameter is internal only and is used after the message is constructed. On .NET Core, the offset is always in the message.

Reply:
I've added code to construct a default message, but this raises a question: all string resources for errors currently include the pattern. So I've changed the constructor of the exception to include the pattern, though that goes against the final design as approved by the api-review team.

We may also consider setting a single default exception text, but that is less useful.

Or we remove the extra public constructor overload completely.

Some of the nested error text resources include additional info (like the char class escape name), that kind of information is lost this way.

Reply:
Now not on phone so I can read properly, sorry about that. We have to implement exactly the shape approved or go back for re-approval if we discover it was a mistake.

As you say many of the error messages we have expect a parameter, but the constructor the API review requested doesn't take that parameter. What message are we supposed to put in the exception for eg "UndefinedBackref" if the message is "Reference to undefined group number {0}." and we don't have the replacement? I think it has to be a generic message, and we won't use this constructor anyway - maybe nobody will.

Also I didn't see that MakeException is private to RegexParser. Given that, it could probably stay there, or if you wanted it on the exception you could have the public approved constructor (which our code wouldn't use) plus an internal constructor.

eg define SR.RegexParseError as `Invalid regular expression pattern. Error '{0}' at offset {1}` or something and
```c#
public RegexParseException(RegexParseError error, int offset) : base(SR.Format(SR.RegexParseError, error, offset)
{
      Error = error;
      Offset = offset;
}
internal RegexParseException(RegexParseError error, int offset, string pattern, string param = null) : base(MakeMessage(error, offset, pattern, param)
{
      Error = error;
      Offset = offset;
}
private static string MakeMessage(RegexParseError error, int offset, string pattern, string param)
{
            string resource = error switch
            {
                RegexParseError.UnterminatedComment => SR.UnterminatedComment),
                //...
            };
            string innerMessage = SR.Format(resource, param);
            return SR.Format(SR.MakeException, pattern, offset, innerMessage);
}
```
Would that work?

Reply:
@danmosemsft, what are your thoughts here?

Reply:
Another thought -- we generally use the "throw helper" pattern in code that might be "hot" because the generated code to construct and throw an exception object is significantly larger than an unconditional call to a throw helper.

It would not be unreasonable (might be neatest) to use that in the regex parser, ie., 

```c#
class RegexParser
{
    // ... parsing
    ThrowRegexParseException(error);

   // elsewhere, ThrowRegexParseException(error, param);

...
   private void ThrowRegexParseException(RegexParseError error) => ThrowRegexParseException(error, null); // overload, not default param, so it's not visible to caller

   private void ThrowRegexParseException(RegexParseError error, string param)
   {
          throw new RegexParseException(error, _currentPos, _pattern, param);
    }
```

Reply:
If it's easier to read, you could rename the resources to match the enum members. So eg`RegexParseError.AlternationHasMalformedCondition => SR.IllegalCondition` becomes `RegexParseError.AlternationHasMalformedCondition => SR.AlternationHasMalformedCondition`

Reply:
It is your preference I think whether to put the "MakeMessage" stuff into ThrowRegexParseException, then the exception itself would just have `internal RegexParseException(RegexParseError error, int offset, string message)`

Reply:
@danmosemsft, as you mentioned elsewhere meanwhile, I've simplified this now with a simple message: `Regular expression parser error '{0}' at offset {1}.`. The first field is filled with the enum value.

Reply:
Crap, somehow my browser didn't refresh everything. Though the code is now much simpler anyway and we're back at the approved api layout.

Reply:
>  I think it has to be a generic message, and we won't use this constructor anyway - maybe nobody will.

Yep, that was my thought on the last commit. Alternatively I could create a special message for each enum value, but I don't really see the point in that.

About the `ThrowHelper` stuff: currently all code paths go through `MakeException`, it'll be a big refactoring to replace all of that.

I had the same thoughts on renaming the resources, but held off on that because (a) they weren't all the same before I started and (b) I wasn't sure what can of worms I would open if I change resource strings (in some other repos that means regenerating all translations). But if you think it's worthwhile, I can still go through that and rename.